### DISCUSSION TITLE
How does `predict_linear` actually work?

### DISCUSSION THREAD

**Discussion by R-Studio** [prometheus/prometheus#11705] [Q&A]
### What did you do?

I want to predict when the volume of Prometheus will be full. Not like an alert "_in 4h your volume will be full_" but more like "_your volume will be full at <Timestamp>_".  

I tried to use the Prometheus function `predict_linear`, but the end of the predicted data are always wrong and weird and maybe this is a bug.

### What did you expect to see?

The predicted data are accurate and I can see when the volume will be full. 

### What did you see instead? Under which circumstances?

I run the following queries and the predicted data looks really wrong/weird and I think this is a bug (_maybe I am using the function wrong, then sorry_).   
Raw data query: `kubelet_volume_stats_available_bytes{namespace="cattle-monitoring-system"} / (1024 * 1024 * 1024)`  
Predicted data query: `predict_linear(kubelet_volume_stats_available_bytes{namespace="cattle-monitoring-system"}[2d], 3600) / (1024 * 1024 * 1024)`  
This is how the graph looks like:  
![image](https://user-images.githubusercontent.com/7498854/201865057-2f7721ef-20e8-4b7c-83bb-c89655d5a6c6.png)

_Only for info: I also tried `holt_winters` but this is not accurate too (but this is not the case for this bug report):_  
 Query: `holt_winters(kubelet_volume_stats_available_bytes{namespace="cattle-monitoring-system"}[2d], 0.5, 0.5) / (1024 * 1024 * 1024)`  
![image](https://user-images.githubusercontent.com/7498854/201869659-804a6ac7-4262-4399-bc46-8c0a6a4c5270.png)

### System information

Linux 5.15.0-52-generic x86_64

### Prometheus version

```text
Version:   2.28.1
Revision:  b0944590a1c9a6b35dc5a696869f75f422b107a1
BuildUser: root@2915dd495090
BuildDate: 20210701-15:20:10
GoVersion: go1.16.5
```

**Comment by beorn7** [comment]
I believe the behavior you are seeing is correct. Your query predicts the value 1 hour (`3600` seconds) into the future, based on the data of the last two days (`2d`). That works well as long as there are 2d of data available. However, once you plot the data for the actual future, your 2d window contains less and less data, until it contains no data at all at 2d into the future. Where you see the very wild changes in the curve, you have very few data points in the 2d window, so the prediction gets very noisy.

**Comment by beorn7** [comment]
As a side note: `predict_linear` is not really suitable to draw a forecast graph in Grafana.

**Comment by R-Studio** [comment]
@beorn7 first thanks for your reply. I don't exactly understand how or you mean do you mean less and less data.

Is there a better PromQL function to draw forecast graphs in Grafana? Where is the limitation of `predict_linear`

**Reply by beorn7** [reply]
> Is there a better PromQL function to draw forecast graphs in Grafana?

Short answer is: No. PromQL is not designed to draw forecasts.

First of all, the forecast capabilities of PromQL are rather limited. You can maybe do some simple things based on the primitives provided by PromQL. The only function really dedicated to forecasting is `predict_linear`, but it is really just doing a regression analysis based on the provided range selector and then draws a line from there. That's cool for an alert based on trending, but not really for sophisticated forecasting. And it's really not at all suitable for drawing that linear forecast. (And even if it were, it would be a really boring graph, just a straight line.)

Generally, if you want to do actual forecasting based on the data in the Prometheus TSDB, you should retrieve the data you want to base your forecast on via the HTTP query API and feed it to a tool that is actually made for forecasting. There are a bunch of people who have done that and published something about it. A recent example is [this PromCon talk](https://promcon.io/2022-munich/talks/tamland-how-gitlabcom-uses-long-/).

**Reply by beorn7** [reply]
> I don't exactly understand how or you mean do you mean less and less data.

So all what `predict_linear` does is it looks at the data in the range selector (in your case the last 2d back from the evaluation time), does a linear regression analysis, takes the slope from there, and uses it to predict as many seconds into the future (from the evaluation time) based on the second argument of `predict_linear` (1h in your case).

If your evaluation time is "now", it looks at the last 2d of data, and extrapolates the slope from there for 1 hour. The result you get is actually not a forecast for "now" but for "now + 1h". And it just works fine (provided you are satisfied with a simple linear "prediction").

If you now graph this expression with Grafana, what happens?

If you graph a data point in the past, let's say 1h in the past, the range selector looks at the last 2d of data as of 1h in the past, i.e. from "now - 2d - 1h" to "now - 1h". It takes the slope, extrapolates it 1h into the "future", i.e. to actually "now". So what Grafana draws for 1h in the past is actually the prediction for "now" but based on the data from "now - 2d - 1h" to "now - 1h". You could compare this to the actual data for "now" to see how accurate the prediction from 1h ago was.

If you graph a data point for 2d into the future, the range selector looks at data from "now - 2d + 2d", i.e. "now", to "now + 2d". The Prometheus TSDB does not (yet) contain data for that time range, so the range selector selects nothing, and `predict_linear` doesn't yield any result.

If you graph a data point for 1d into the future,  the range selector looks at data from "now - 2d + 1d", i.e. "now - 1d", to "now + 1d". Now we do find _some_ data in the TSDB for that range, but only in its first half ("now -1d" to "now"). So `predict_linear` calculates a sloped based on only 1d of data, and uses it to predict the value for 1d+1h in the future.

What you probably want is to _always_ look at the last 2d before "now", and take that slope to predict all the future values. But that's not how PromQL works. (And as said, the predicted graph would just be a straight line anyway.)

**Comment by beorn7** [comment]
I have converted this into a discussion as I'm fairly certain that the reported behavior isn't a bug in `predict_linear` but the expected and desired behavior.
